prow/git/git.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package git provides a client to plugins that can do git operations.
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+)
+
+const github = "https://github.com"
+
+// Client can clone repos. It keeps a local cache, so successive clones of the
+// same repo should be quick. Create with NewClient. Be sure to clean it up.
+type Client struct {
+	dir  string
+	git  string
+	base string
+
+	// The mutex protects repoLocks which protect individual repos. This is
+	// necessary because Clone calls for the same repo are racy. Rather than
+	// one lock for all repos, use a lock per repo.
+	rlm       sync.Mutex
+	repoLocks map[string]*sync.Mutex
+}
+
+// Clean removes the local repo cache. The Client is unusable after calling.
+func (c *Client) Clean() error {
+	return os.RemoveAll(c.dir)
+}
+
+// NewClient returns a client that talks to GitHub. It will fail if git is not
+// in the PATH.
+func NewClient() (*Client, error) {
+	g, err := exec.LookPath("git")
+	if err != nil {
+		return nil, err
+	}
+	t, err := ioutil.TempDir("", "git")
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		dir:       t,
+		git:       g,
+		base:      github,
+		repoLocks: make(map[string]*sync.Mutex),
+	}, nil
+}
+
+func (c *Client) lockRepo(repo string) {
+	c.rlm.Lock()
+	defer c.rlm.Unlock()
+	if _, ok := c.repoLocks[repo]; !ok {
+		c.repoLocks[repo] = &sync.Mutex{}
+	}
+	c.repoLocks[repo].Lock()
+}
+
+func (c *Client) unlockRepo(repo string) {
+	c.rlm.Lock()
+	defer c.rlm.Unlock()
+	c.repoLocks[repo].Unlock()
```

Comment:
I added a comment that we should use `lockRepo` and `unlockRepo` to lock and unlock these. If we do so then there is no way to reach this point without the repo lock existing.